src/Elf.cpp

```diff
@@ -0,0 +1,989 @@
+#include "Elf.h"
+#include "Debug.h"
+#include "Util.h"
+#include "Error.h"
+
+#include <algorithm>
+#include <map>
+#include <array>
+#include <memory>
+#include <iomanip>
+
+namespace Halide {
+namespace Internal {
+namespace Elf {
+
+namespace {
+
+// http://www.skyfree.org/linux/references/ELF_Format.pdf
+
+enum : uint32_t {
+    PT_NULL = 0,
+    PT_LOAD = 1,
+    PT_DYNAMIC = 2,
+    PT_INTERP = 3,
+    PT_NOTE = 4,
+    PT_SHLIB = 5,
+    PT_PHDR = 6,
+    PT_LOPROC = 0x70000000,
+    PT_HIPROC = 0x7fffffff,
+};
+
+enum : uint32_t {
+    PF_X = 1,
+    PF_W = 2,
+    PF_R = 4,
+    PF_MASKOS = 0x0ff00000,
+    PF_MASKPROC = 0xf0000000,
+};
+
+enum : uint32_t {
+    DT_NULL = 0,
+    DT_NEEDED = 1,
+    DT_PLTRELSZ = 2,
+    DT_PLTGOT = 3,
+    DT_HASH = 4,
+    DT_STRTAB = 5,
+    DT_SYMTAB = 6,
+    DT_RELA = 7,
+    DT_RELASZ = 8,
+    DT_RELAENT = 9,
+    DT_STRSZ = 10,
+    DT_SYMENT = 11,
+    DT_INIT = 12,
+    DT_FINI = 13,
+    DT_SONAME = 14,
+    DT_RPATH = 15,
+    DT_SYMBOLIC = 16,
+    DT_REL = 17,
+    DT_RELSZ = 18,
+    DT_RELENT = 19,
+    DT_PLTREL = 20,
+    DT_DEBUG = 21,
+    DT_TEXTREL = 22,
+    DT_JMPREL = 23,
+    DT_LOPROC = 0x70000000,
+    DT_HIPROC = 0x7fffffff,
+};
+
+enum : uint32_t {
+    STN_UNDEF = 0
+};
+
+static const char elf_magic[] = { 0x7f, 'E', 'L', 'F' };
+
+template <int bits>
+struct Types;
+
+template <>
+struct Types<32> {
+    typedef uint32_t addr_t;
+    typedef int32_t addr_off_t;
+};
+
+template <typename T>
+struct Ehdr {
+    typedef typename T::addr_t addr_t;
+    typedef typename T::addr_off_t addr_off_t;
+
+    uint8_t e_ident[16];
+    uint16_t e_type;
+    uint16_t e_machine;
+    uint32_t e_version;
+    addr_t e_entry;
+    addr_t e_phoff;
+    addr_t e_shoff;
+    uint32_t e_flags;
+    uint16_t e_ehsize;
+    uint16_t e_phentsize;
+    uint16_t e_phnum;
+    uint16_t e_shentsize;
+    uint16_t e_shnum;
+    uint16_t e_shstrndx;
+};
+
+
+template <typename T>
+struct Phdr {
+    typedef typename T::addr_t addr_t;
+    typedef typename T::addr_off_t addr_off_t;
+
+    uint32_t p_type;
+    uint32_t p_offset;
+    addr_t p_vaddr;
+    addr_t p_paddr;
+    uint32_t p_filesz;
+    uint32_t p_memsz;
+    uint32_t p_flags;
+    uint32_t p_align;
+};
+
+template <typename T>
+struct Shdr {
+    typedef typename T::addr_t addr_t;
+    typedef typename T::addr_off_t addr_off_t;
+
+    uint32_t sh_name;
+    uint32_t sh_type;
+    addr_t sh_flags;
+    addr_t sh_addr;
+    addr_t sh_offset;
+    addr_t sh_size;
+    uint32_t sh_link;
+    uint32_t sh_info;
+    addr_t sh_addralign;
+    addr_t sh_entsize;
```

Comment:
Hmm, it looks like this is a type that is the same as uintptr_t (which is addr_t), looking at Elf32_shdr and Elf64_shdr here: https://github.com/torvalds/linux/blob/master/include/uapi/linux/elf.h#L312

Still, I will clean this up a bit.

We can't include the linux header because this needs to build on windows.